### PR TITLE
sparseml integration

### LINKED ISSUE #196: Sparseml Integration
## 🚀 Feature

Add option to use sparseml. Example implementation found here: [Google Colab link](https://colab.research.google.com/drive/1cIQRC8zKK_c8_Y65CgcNxl-C4oaGi2jg?usp=sharing)

### Motivation

There is currently no option to use pruning techniques from sparseml out-of-the-box

### Pitch

I will make a pull request to add this option to the hydra config. I've already forked a version of the lightning-tranformers library. [link](https://github.com/mathemusician/lightning-transformers)

Here's how it will be added on the hydra CLI:

```bash
trainer=sparseml
```

Passing this into the trainer means it will automatically use ddp. How convenient! Sparseml also uses a special conversion to log weights and such, and I've also implemented this:

```bash
+trainer/logger=sparsewandb
```

It is also available as a callback for those who want to train with CPU.
```
+trainer/callback=sparseml
```

Sparseml barely supports transformers at the moment, so I've had to make a workaround for their exporter. BERT and other BERT-like models output a ModelOutput which tell the exporter there will be two outputs. But sometimes, there's only one. I've just forced the exporter to treat is all as one output for now. I may open a pull request at [sparseml](https://github.com/neuralmagic/sparseml/) to handle transformer outputs.

The ```RECIPE_PATH``` and ```MODELS_PATH```, paths to the recipe yaml and models folder, are passed in as environment variables. I wasn't able to find a way to around this since hydra overwrites added configs after starting the training loop. Maybe there's a better way of doing this.

### Alternatives

I haven't thought much about this, but I'll add in a few good alternatives once I find some.

### Additional context

This is my first time to make a pull request to a rather large library, so don't be afraid to critique. I need the feedback. I may also need help understanding how to get the "fit" stage works differently from the  "train" stage. I'm running ```training.run_test_after_fit=False```  becuase the fit stage doesn't work. Training works just fine, however.

### PR DESCRIPTION
**Pull request by mathemusician**
Closes #196 

Add option to use sparseml. This is still a little rough around the edges, but I welcome the feedback! Implementation on [Google Colab](https://colab.research.google.com/drive/1cIQRC8zKK_c8_Y65CgcNxl-C4oaGi2jg?usp=sharing).

Still needs:

1) Implementation of ```MODELS_PATH``` and ```RECIPE_PATH``` through Hydra instead of environment variable
2) "test" stage needs to work, only training and evaluation work at the moment
3) stylistic coding changes

Will continue to add more changes after my reactions engineering exam tomorrow. Feel free to give me pointers or links to tutorials if you see any changes I need to make.

### CODE REVIEW

**Comment by mathemusician** [comment]
Wow, that's a lot of red. I think I may be able to get rid of the dependency to pytorch-lightning-bolts, which isn't even listed as a dependency yet. Also, @SeanNaren I've tested sparseml with torch 1.9, and it works!? I'll do a more thorough test after my test tomorrow and add the appropriate changes.

File: lightning_transformers/core/callback.py [hunk 1]
```diff
@@ -11,11 +11,123 @@
...
 from pytorch_lightning import Callback
+from pytorch_lightning.loggers import WandbLogger
 from pytorch_lightning.utilities import rank_zero_info
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from sparseml.pytorch.utils import ModuleExporter
+from sparseml.pytorch.utils.logger import WANDBLogger
+from torch import Tensor
+
+
+class LightningBoltsSparseMLCallback(SparseMLCallback):
```
Comment (Borda) [inline lightning_transformers/core/callback.py:34]:
```suggestion
class TranformerSparseMLCallback(SparseMLCallback):
```

File: lightning_transformers/core/callback.py [hunk 2]
```diff
@@ -11,11 +11,123 @@
...
+                    exporter.export_onnx(
+                        sample_batch=one_sample_input,
+                        name="small_model.onnx",
+                        convert_qat=True,
+                        export_params=False,
+                        **kwargs,
+                    )
+                    onnx_file = os.path.join(output_dir, "model.onnx")
+
+                except Exception:
```
Comment (Borda) [inline lightning_transformers/core/callback.py:105]: pls be more specific what exception is expected here

File: lightning_transformers/core/loggers.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import time
+from typing import Dict, Optional, Union
+
+from pytorch_lightning.loggers import WandbLogger
+from sparseml.pytorch.utils.logger import LambdaLogger
+
+
+class WANDBLogger(WandbLogger):
```
Comment (Borda) [inline lightning_transformers/core/loggers.py:8]: WANDB is not an abbreviation, so pls follow only the first letter convention

**Comment by SeanNaren** [comment]
Thanks @mathemusician I'll be able to assist next week :)

File: lightning_transformers/core/callback.py (hunk 1 above)
Comment (mathemusician) [reply lightning_transformers/core/callback.py:34]: 👍 I'll commit that with the correct spelling of Transformer

File: lightning_transformers/core/callback.py (hunk 2 above)
Comment (mathemusician) [reply lightning_transformers/core/callback.py:105]: changed to RunTimeError

File: lightning_transformers/core/loggers.py (hunk 3 above)
Comment (mathemusician) [reply lightning_transformers/core/loggers.py:8]: Interesting, the convention actually seems to be to refer to weights & biases as wandb, but I'll change it to WABLogger to follow first letter convention

**Comment by mathemusician** [comment]
Figured out why the "test" stage doesn't work. The labels given to bert during the test stage are ```'labels': tensor([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1])```. When bert tries to calculate cross-entropy loss, it fails immediately. In order to fix this, I added a check in task/nlp/text_classification/model.py to check for -1 in labels; if there is -1 it replaces labels with None. Test stage now works splendidly.

**Comment by mathemusician** [comment]
Also, I figured out how to change the recipe path and output directory after instantiation by hydra:

```
++trainer.callbacks.output_dir=/content/MODELS            \
++trainer.callbacks.recipe_path=/content/recipe.yaml     \
```

I'm getting rid of the dependency on environment variables

File: conf/trainer/sparseml.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+defaults:
+  - default # inherit from default trainer conf
+  - callbacks: sparseml
+  - logger: sparsewandb
```
Comment (SeanNaren) [inline conf/trainer/sparseml.yaml:4]: hey @mathemusician should we make this logger the default? seems unnecessary but let me know :)

File: conf/trainer/logger/sparsewandb.yaml [hunk 5]
```diff
@@ -0,0 +1 @@
+_target_: lightning_transformers.core.loggers.WABLogger
```
Comment (SeanNaren) [inline conf/trainer/logger/sparsewandb.yaml:1]: Is this logger a necessity? Could we add this in a separate PR? why is it needed for SparseML?

**Comment by SeanNaren** [comment]
PR looks good! Any chance you could add a quick test for one of the tasks to ensure the callback works end to end?

File: conf/trainer/logger/sparsewandb.yaml (hunk 5 above)
Comment (mathemusician) [reply conf/trainer/logger/sparsewandb.yaml:1]: SparseML calls special functions (_lambda_func, log_scalar, and log_scalars) to the logger. Why? It looks like they needed it to convert certain arguments to JSON-compatible format. Instead of making a wrapper for loggers, what they should have done instead is convert those arguments before feeding them into the logger. This makes it impossible to use other loggers directly without making an entirely new wrapper.  I'm thinking about opening a PR at the sparseml repository to get their code to talk normally to a logger.

To prevent the user from accidentally using wandb from default, I've included it as the default. Until someone (probably me) opens a PR at SparseML, this is unfortunately what we have to work with.

File: conf/trainer/sparseml.yaml (hunk 4 above)
Comment (mathemusician) [reply conf/trainer/sparseml.yaml:4]: I answer this question in more detail below

**Comment by mathemusician** [comment]
> PR looks good! Any chance you could add a quick test for one of the tasks to ensure the callback works end to end?

I'll try, it's my first time writing tests. Gotta start somewhere.

File: conf/trainer/logger/sparsewandb.yaml (hunk 5 above)
Comment (SeanNaren) [reply conf/trainer/logger/sparsewandb.yaml:1]: Makes sense, thanks @mathemusician :)

**Comment by SeanNaren** [comment]
> > PR looks good! Any chance you could add a quick test for one of the tasks to ensure the callback works end to end?
> 
> I'll try, it's my first time writing tests. Gotta start somewhere.

let me know if you get stuck, more than happy to help out! Code is really well thought out so far, so I have no hesitations that you can't do it :D

**Comment by mathemusician** [comment]
The logger and callback tests need certain imports that aren't in requirements.txt. But I wasn't sure if we wanted to add more import statements. The logger test needs wandb; and the sparseml callback test needs pytorch 1.7-1.8.

**Comment by mathemusician** [comment]
Should I add wandb to requirements.txt? Also, I'm going to update the docs with the different ways to use sparseml soon

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
Thanks for your hard work @mathemusician hopefully tests should pass now, and we can merge :)
